### ISSUE TITLE
Provide PSM chainStorage data to Bulldozer bootstrap

### ISSUE THREAD

**Issue by dckc** [Agoric/Agoric#7156]
## What is the Problem Being Solved?

In order to, for example, restore PSM state from chainStorage (#6645), the post-bulldozer VM has to somehow access the pre-bulldozer chainStorage state such as the list of PSM assets.

## Description of the Design

???

Options include:

 - add a `getValue()` method to the JS chainStorage API
    - looks relatively straightforward (it's already in the [go API](https://github.com/Agoric/agoric-sdk/issues/6645#issuecomment-1405773095)).
    - But in a [Jan 26 discussion with @warner ](https://github.com/Agoric/agoric-sdk/issues/6645#issuecomment-1405773095), @warner expressed a desire to avoid reading from SwingSet devices (for reasons that I didn't manage to capture well).
  - put the whole pre-bulldozer state in vatParameters
    - those vatParameters get persisted in the kernel DB, and now replicated back into vstorage

**Comment by warner** [comment]
The data we're talking about is pretty limited:
* the PSM contracts have roughly 6 parameters to track, I think it's a balance and a mint limit for each denomination
* the smart-wallet needs a list of addresses, one per account that was provisioned in the pismo chain
  * it does not need their balances, just a Set
  * if/when someone submits a smart-wallet action, their address will be compared against this set to see if they must provision an account first, or if they're already set up

In today's kernel meeting, we agreed:
* `launch-chain.js` (the JS side of cosmic-swingset) will be changed to, at startup time, ask the Golang side for the contents of certain IAVL keys
* it will then format that data into a JSON-shaped blob and deliver it to the kernel in the `bootstrapArgs` parameter (passed to `initializeSwingset`)
* swingset delivers that data to vat-bootstrap (as `bootstrap(vats, devices, bootstrapArgs)`)
* vat-bootstrap will be changed to disseminate that data to the PSM/wallet vats which need it

@mhofman will mostly be doing the launch-chain.js side (this ticket), and @dckc will mostly be doing the vat-bootstrap side (#6645).

This will result in some specialized knowledge living in an unusual place. The particular set of IAVL keys that hold the PSM/wallet data is a function of the behavior of the pismo-era vats, and the IAVL root key that chain-storage starts from (I think it starts with `swingset.` or `swingset.published.` or something). The format of the contents of those keys is defined by those old vats too. But `launch-chain.js` will be required to have a list of those keys, or some code that knows how to find them, even though `launch-chain.js` was not involved in putting data there originally.

The alternative, which we rejected, would be for the PSM/smart-wallet vats to get chain-storage *read* access, whereupon at first launch they could query for data written by their predecessor, and then populate their initial (durable) tables from the results. That would separate the concerns better (new PSM vat is solely responsible for recovering state from the old PSM vat), but would require chain-storage reads, which we believe would interfere with our plans to parallelize vat execution.

Basically if a vat can ever read (synchronously) from a device, the worker hosting that vat would have to stop its execution, send a read request to the kernel, and wait for its response before it could continue execution. The kernel would receive that read request at a time that is uncoordinated with other activity, which looks like a timing race, and a threat to deterministic execution. The parallel execution design will be a lot simpler if we don't have to accomodate this case, and stick to only doing device *writes* (which can be safely deferred until the end of the vat delivery).

**Comment by mhofman** [comment]
To avoid too much (any) economy logic in cosmic-swingset, it would be preferable for the config to list the keys/path that need to be read/exported and included in the bootstrap args.

@dckc do we have an idea of the kind of vstorage read operations we need? read specific paths? export a whole subtree? dump a list of children and/or entries?

**Comment by dckc** [comment]
If the swingset config can specify a list of roots (`['published.psm', 'published.wallet']` for example) and get an `Object.entries()` style dump of those subtrees, that would probably work nicely. And no need for history. Just current/latest values.

**Comment by mhofman** [comment]
I would imagine that an export of those subtrees will be more data than we need.

Also from what I recall the "history" is an application concern since it's an append to the same vstorage paths, so you would indeed get the history. Maybe we should look at an export of these subtrees and see what we can do from it.

**Comment by turadg** [comment]
assigning to @gibson042 who volunteered for this in a chat thread

**Comment by warner** [comment]
cc @FUDCo

**Comment by dckc** [comment]
FYI: I started prototyping by grabbing data straight off mainnet:

https://github.com/Agoric/agoric-sdk/blob/1ff889104d1c790cb08c17ee5c8c9e71f85b24bd/packages/inter-protocol/test/psm/test-psm.js#L712-L723

**Comment by gibson042** [comment]
@dckc I'm also planning to use that shape (an array of [key: string, value: string] entries). But you do need a full subtree, right?

**Comment by dckc** [comment]
> you do need a full subtree, right?

Yes; in particular, for `published.wallet`. But it's also convenient for `published.psm.IST`.

I'm not sure whether I need `published.agoricNames` or `published.agoricNames.brand`... that part of my prototype might not survive.
